presto-hive/src/main/java/com/facebook/presto/hive/HivePageSource.java

```diff
@@ -424,6 +440,163 @@ public Block apply(Block block)
         }
     }
 
+    private static class ListCoercer
+            implements Function<Block, Block>
+    {
+        private final Type toType;
+        private final Function<Block, Block> elementCoercer;
+
+        public ListCoercer(TypeManager typeManager, HiveType fromHiveType, HiveType toHiveType)
+        {
+            requireNonNull(typeManager, "typeManage is null");
+            requireNonNull(fromHiveType, "fromHiveType is null");
+            this.toType = requireNonNull(toHiveType, "toHiveType is null").getType(typeManager);
+            HiveType fromElementHiveType = HiveType.valueOf(((ListTypeInfo) fromHiveType.getTypeInfo()).getListElementTypeInfo().getTypeName());
+            HiveType toElementHiveType = HiveType.valueOf(((ListTypeInfo) toHiveType.getTypeInfo()).getListElementTypeInfo().getTypeName());
+            this.elementCoercer = fromElementHiveType.equals(toElementHiveType) ? null : createCoercer(typeManager, fromElementHiveType, toElementHiveType);
+        }
+
+        @Override
+        public Block apply(Block block)
+        {
+            if (elementCoercer == null) {
+                return block;
+            }
+            BlockBuilder blockBuilder = toType.createBlockBuilder(new BlockBuilderStatus(), block.getPositionCount());
```

Comment:
In all 3, do not use `BlockBuilder`.

Instead, new `new ArrayBlock`, `mapType.createBlockFromKeyValue`, and `new RowBlock`. This is the same way they are done in XxxStreamReader in presto-orc.

This is likely much easier to implement as well.

Performance is critical HivePageSource as ORC, Parquet, DWRF goes through this path. In a typical Presto set up, more than 80% of time is spent in TableScanOperator and ScanFilterProjectOperator.

Reply:
I tried to refactor `ListCoercer`. But I realize it's not visible to get the offsets of coerced element block. Any idea?